File: src/main/java/io/github/mb1lal/playwright/ui/pages/heroku/MultipleWindowsPage.java
package io.github.mb1lal.playwright.ui.pages.herokuapp;

import io.github.mb1lal.playwright.base.TestContext;
import io.github.mb1lal.playwright.ui.pages.BasePage;
import com.microsoft.playwright.Page;
import lombok.extern.slf4j.Slf4j;

/**
 * Multiple Windows page ({@code /windows}): the link opens a new tab.
 */
@Slf4j
public class MultipleWindowsPage extends BasePage {

    public MultipleWindowsPage(Page page, TestContext testContext) {
        super(page, testContext);
    }

    /** Click "Click Here" and return the newly opened page. */
    public Page openNewWindow() {
        Page popup = page.waitForPopup(() ->
                click(page.locator("text=Click Here").first(), "Click Here link"));
        popup.waitForLoadState();
        log.info("New window opened at {}", popup.url());
        return popup;
    }
}

File: typescript/src/pages/heroku/multiple-windows.page.ts
import { Page } from '@playwright/test';
import { BasePage } from '../base.page';

/** Multiple Windows page (/windows): the link opens a new tab. */
export class MultipleWindowsPage extends BasePage {
  /** Click "Click Here" and return the newly opened page. */
  async openNewWindow(): Promise<Page> {
    const [popup] = await Promise.all([
      this.page.waitForEvent('popup'),
      this.page.locator('text=Click Here').first().click(),
    ]);
    await popup.waitForLoadState();
    return popup;
  }
}
